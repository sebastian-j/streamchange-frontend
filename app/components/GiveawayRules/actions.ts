/*
 *
 * GiveawayRules actions
 *
 */
import { action } from 'typesafe-actions';
import { User } from '../UserList/types';
import ActionTypes from './constants';

export const changeKeyword = (keyword: string) =>
  action(ActionTypes.CHANGE_KEYWORD, keyword);

export const changePreWinner = (preWinner: User | null) =>
  action(ActionTypes.CHANGE_PREWINNER, preWinner);

export const changePrize = (prize: string) =>
  action(ActionTypes.CHANGE_PRIZE, prize);

export const changeRequirement = (requirement: number) =>
  action(ActionTypes.CHANGE_PARTICIPANT_REQUIREMENT, requirement);
