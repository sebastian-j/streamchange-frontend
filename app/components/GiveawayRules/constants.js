/*
 *
 * GiveawayRules constants
 *
 */

export const CHANGE_KEYWORD = 'app/GiveawayRules/CHANGE_KEYWORD';
export const CHANGE_PREWINNER = 'app/GiveawayRules/CHANGE_PREWINNER';
export const CHANGE_PRIZE = 'app/GiveawayRules/CHANGE_PRIZE';
export const CHANGE_PARTICIPANT_REQUIREMENT =
  'app/GiveawayRules/CHANGE_PARTICIPANT_REQUIREMENT';
