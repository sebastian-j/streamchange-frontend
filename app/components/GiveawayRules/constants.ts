/*
 *
 * GiveawayRules constants
 *
 */

enum ActionTypes {
  CHANGE_KEYWORD = 'app/GiveawayRules/CHANGE_KEYWORD',
  CHANGE_PREWINNER = 'app/GiveawayRules/CHANGE_PREWINNER',
  CHANGE_PRIZE = 'app/GiveawayRules/CHANGE_PRIZE',
  CHANGE_PARTICIPANT_REQUIREMENT = 'app/GiveawayRules/CHANGE_PARTICIPANT_REQUIREMENT',
}

export default ActionTypes;
