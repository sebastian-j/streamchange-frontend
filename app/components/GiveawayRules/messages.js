/*
 * GiveawayRules Messages
 *
 * This contains all the text for the GiveawayRules component.
 */
import { defineMessages } from 'react-intl';

export const scope = 'streamchange.components.GiveawayRules';

export default defineMessages({
  panelTitle: {
    id: `${scope}.title`,
    defaultMessage: 'Giveaway rules',
  },
  allViewers: {
    id: `${scope}.forAll`,
    defaultMessage: 'All viewers',
  },
  sponsors: {
    id: `${scope}.forSponsors`,
    defaultMessage: 'Sponsors only',
  },
  userTypeLabel: {
    id: `${scope}.userTypeLabel`,
    defaultMessage: 'Who can participate in raffle',
  },
  keyword: {
    id: `${scope}.KeywordInput.label`,
    defaultMessage: 'Keyword',
  },
  keywordVisibility: {
    id: `${scope}.KeywordInput.visibility`,
    defaultMessage: 'Toggle keyword visibility',
  },
  prize: {
    id: `${scope}.prize`,
    defaultMessage: 'Prize (optional)',
  },
});
