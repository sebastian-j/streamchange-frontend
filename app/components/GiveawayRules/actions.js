/*
 *
 * GiveawayRules actions
 *
 */

import { CHANGE_KEYWORD, CHANGE_PREWINNER, CHANGE_PRIZE } from './constants';

export function changeKeyword(str) {
  return {
    type: CHANGE_KEYWORD,
    keyword: str,
  };
}

export function changePreWinner(w) {
  return {
    type: CHANGE_PREWINNER,
    preWinner: w,
  };
}

export function changePrize(str) {
  return {
    type: CHANGE_PRIZE,
    prize: str,
  };
}
