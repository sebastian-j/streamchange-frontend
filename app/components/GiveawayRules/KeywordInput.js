import React, { useState, useEffect } from 'react';
import styled from 'styled-components';
import InputAdornment from '@material-ui/core/InputAdornment';
import Tooltip from '@material-ui/core/Tooltip';
import { FormattedMessage } from 'react-intl';

import messages from './messages';
import StyledTextField from '../StyledTextField';

const ToggleVisibilityBtn = styled.button`
  background-color: transparent;
  border: none;
  color: ${props => props.theme.staticTextColor};
  outline: none;
  &:hover {
    color: #0094ff;
  }
  & * {
    pointer-events: none;
  }
`;

const KeywordInput = () => {
  const [keyword, setKeyword] = useState('');
  const [visible, setVisible] = useState(true);
  const [error, setError] = useState(null);

  useEffect(() => {
    let storedKeyword = localStorage.getItem('keyword');
    if (storedKeyword === null) {
      storedKeyword = '';
    }
    setKeyword(storedKeyword);
  }, []);

  const handleInputChange = event => {
    if (event.target.value.includes(localStorage.getItem('gv-abortCommand'))) {
      setError('Komenda na dołączenie nie może zawierać komendy rezygnacji.');
      return;
    }
    setError(null);
    setKeyword(event.target.value);
    localStorage.setItem('keyword', event.target.value);
  };
  const handleToggleButton = () => setVisible(prevState => !prevState);

  return (
    <FormattedMessage {...messages.keyword}>
      {label => (
        <StyledTextField
          autoFocus
          error={!!error}
          margin="dense"
          name="keyword"
          onChange={handleInputChange}
          id="keyword"
          label={label}
          type={visible ? 'text' : 'password'}
          value={keyword}
          helperText={error}
          fullWidth
          InputProps={{
            endAdornment: (
              <InputAdornment position="end">
                <FormattedMessage {...messages.keywordVisibility}>
                  {title => (
                    <Tooltip title={title} aria-label="add">
                      <ToggleVisibilityBtn
                        aria-label="toggle keyword visibility"
                        name="showKeyword"
                        onClick={handleToggleButton}
                        type="button"
                      >
                        {visible ? (
                          <svg width="24" height="24" viewBox="0 0 24 24">
                            <path d="M0 0h24v24H0z" fill="none" />
                            <path d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z" />
                          </svg>
                        ) : (
                          <svg width="24" height="24" viewBox="0 0 24 24">
                            <path
                              d="M0 0h24v24H0zm0 0h24v24H0zm0 0h24v24H0zm0 0h24v24H0z"
                              fill="none"
                            />
                            <path d="M12 7c2.76 0 5 2.24 5 5 0 .65-.13 1.26-.36 1.83l2.92 2.92c1.51-1.26 2.7-2.89 3.43-4.75-1.73-4.39-6-7.5-11-7.5-1.4 0-2.74.25-3.98.7l2.16 2.16C10.74 7.13 11.35 7 12 7zM2 4.27l2.28 2.28.46.46C3.08 8.3 1.78 10.02 1 12c1.73 4.39 6 7.5 11 7.5 1.55 0 3.03-.3 4.38-.84l.42.42L19.73 22 21 20.73 3.27 3 2 4.27zM7.53 9.8l1.55 1.55c-.05.21-.08.43-.08.65 0 1.66 1.34 3 3 3 .22 0 .44-.03.65-.08l1.55 1.55c-.67.33-1.41.53-2.2.53-2.76 0-5-2.24-5-5 0-.79.2-1.53.53-2.2zm4.31-.78l3.15 3.15.02-.16c0-1.66-1.34-3-3-3l-.17.01z" />
                          </svg>
                        )}
                      </ToggleVisibilityBtn>
                    </Tooltip>
                  )}
                </FormattedMessage>
              </InputAdornment>
            ),
          }}
        />
      )}
    </FormattedMessage>
  );
};

export default KeywordInput;
