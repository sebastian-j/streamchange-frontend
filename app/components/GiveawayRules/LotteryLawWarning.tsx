import React, {useEffect, useState} from 'react';
import { FormattedMessage } from 'react-intl';
import Button from '@mui/material/Button';
import Dialog from '@mui/material/Dialog';
import DialogActions from '@mui/material/DialogActions';
import DialogContent from '@mui/material/DialogContent';
import DialogTitle from '@mui/material/DialogTitle';

import messages from './messages';

interface Props {
  open: boolean;
}

const LotteryLawWarning = ({ open }: Props) => {
  const [isOpenDialog, setIsOpenDialog] = useState<boolean>(false);

  const closeDialog = () => {
    setIsOpenDialog(false);
  };

  const openDocs = () => {
    window.open('https://www.streamchange.pl/docs/lottery-legal-issues','_blank');
  };

  useEffect(() => {
    setIsOpenDialog(open);
  }, [open]);

  return (
    <Dialog
      open={isOpenDialog}
      onClose={closeDialog}
      aria-labelledby="form-dialog-title"
    >
      <DialogTitle id="form-dialog-title">
        <FormattedMessage {...messages.lawTitle} />
      </DialogTitle>
      <DialogContent>
        <span>
          <FormattedMessage {...messages.lawContent} />
        </span>
      </DialogContent>
      <DialogActions>
        <Button onClick={openDocs} color="primary">
          <FormattedMessage {...messages.lawOpenDocs} />
        </Button>
        <Button onClick={closeDialog} color="primary">
          <FormattedMessage {...messages.lawClose} />
        </Button>
      </DialogActions>
    </Dialog>
  );
};

export default LotteryLawWarning;
