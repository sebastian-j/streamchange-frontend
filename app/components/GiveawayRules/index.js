import React from 'react';
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import { connect } from 'react-redux';
import { FormattedMessage } from 'react-intl';
import InputLabel from '@material-ui/core/InputLabel';
import MenuItem from '@material-ui/core/MenuItem';
import Select from '@material-ui/core/Select';

import messages from './messages';
import {
  makeSelectGiveawayPrize,
  makeSelectGiveawayRequirement,
} from './selectors';
import { changePrize, changeRequirement } from './actions';
import AdFrame from '../AdFrame';
import Panel from '../Panel';
import PanelTitle from '../Panel/PanelTitle';
import StyledTextField from '../StyledTextField';
import StyledFormControl from '../StyledTextField/StyledFormControl';
import KeywordInput from './KeywordInput';
import RaffleWrapper from '../RaffleWrapper';
import WinnerView from '../WinnerView';

export class GiveawayRules extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      winnerId: null,
    };
    this.handleInputValueChange = this.handleInputValueChange.bind(this);
    this.winHandler = this.winHandler.bind(this);
  }

  handleInputValueChange(event) {
    const { target } = event;
    const value = target.type === 'checkbox' ? target.checked : target.value;
    const { name } = target;

    if (value.length < 140) {
      this.setState({
        [name]: value,
      });
      localStorage.setItem(`gv-${name}`, value);
    }
  }

  winHandler(id) {
    this.setState({
      winnerId: id,
    });
  }

  render() {
    if (this.state.winnerId) {
      return (
        <WinnerView
          apiKey={this.props.apiKey}
          id={this.state.winnerId}
          prize={this.props.prize}
          onClose={() => this.setState({ winnerId: null })}
        />
      );
    }
    return (
      <Panel>
        <PanelTitle>
          <FormattedMessage {...messages.panelTitle} />
        </PanelTitle>
        <StyledFormControl margin="normal">
          <InputLabel id="user-type-select">
            <FormattedMessage {...messages.userTypeLabel} />
          </InputLabel>
          <Select
            onChange={(event) => this.props.changeReq(event.target.value)}
            value={this.props.requirement}
          >
            <MenuItem value={0}>
              <FormattedMessage {...messages.allViewers} />
            </MenuItem>
            <MenuItem value={1}>
              <FormattedMessage {...messages.sponsors} />
            </MenuItem>
          </Select>
        </StyledFormControl>
        <FormattedMessage {...messages.prize}>
          {(label) => (
            <StyledTextField
              autoFocus
              margin="dense"
              name="prize"
              onChange={(event) => {
                this.props.changePrize(event.target.value);
              }}
              label={label}
              type="text"
              value={this.props.prize}
              fullWidth
            />
          )}
        </FormattedMessage>
        <KeywordInput />
        <RaffleWrapper onWin={this.winHandler} />
        <AdFrame />
      </Panel>
    );
  }
}

GiveawayRules.propTypes = {
  apiKey: PropTypes.string.isRequired,
  changePrize: PropTypes.func.isRequired,
  changeReq: PropTypes.func.isRequired,
  prize: PropTypes.string,
  requirement: PropTypes.number,
};

const mapStateToProps = createStructuredSelector({
  prize: makeSelectGiveawayPrize(),
  requirement: makeSelectGiveawayRequirement(),
});

export function mapDispatchToProps(dispatch) {
  return {
    changePrize: (a) => dispatch(changePrize(a)),
    changeReq: (r) => dispatch(changeRequirement(r)),
    dispatch,
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(GiveawayRules);
