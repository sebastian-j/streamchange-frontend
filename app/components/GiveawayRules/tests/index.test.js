import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import GiveawayRules from '../index';

const renderer = new ShallowRenderer();

describe('<GiveawayRules />', () => {
  it('should render and match the snapshot', () => {
    renderer.render(<GiveawayRules apiKey="key" channelId="id" />);
    const renderedOutput = renderer.getRenderOutput();
    expect(renderedOutput).toMatchSnapshot();
  });
});
