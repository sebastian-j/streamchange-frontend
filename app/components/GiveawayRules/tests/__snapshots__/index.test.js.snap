// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GiveawayRules /> should render and match the snapshot 1`] = `
<ForwardRef>
  <ForwardRef>
    <FormattedMessage
      defaultMessage="Giveaway rules"
      id="streamchange.components.GiveawayRules.title"
      values={Object {}}
    />
  </ForwardRef>
  <ForwardRef
    active={true}
    name="forMods"
    onClick={[Function]}
    type="button"
  >
    <FormattedMessage
      defaultMessage="Moderators"
      id="streamchange.components.GiveawayRules.moderators"
      values={Object {}}
    />
  </ForwardRef>
  <ForwardRef
    active={true}
    name="forSponsors"
    onClick={[Function]}
    type="button"
  >
    <FormattedMessage
      defaultMessage="Sponsors"
      id="streamchange.components.GiveawayRules.sponsors"
      values={Object {}}
    />
  </ForwardRef>
  <ForwardRef
    active={true}
    name="forRegulars"
    onClick={[Function]}
    type="button"
  >
    <FormattedMessage
      defaultMessage="Regulars"
      id="streamchange.components.GiveawayRules.regulars"
      values={Object {}}
    />
  </ForwardRef>
  <FormattedMessage
    defaultMessage="Prize (optional)"
    id="streamchange.components.GiveawayRules.prize"
    values={Object {}}
  >
    [Function]
  </FormattedMessage>
  <KeywordInput />
  <Memo(Connect(RaffleWrapper))
    onWin={[Function]}
  />
</ForwardRef>
`;
